File: sort/bucket_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void selectSort(vector<int>& nums) {
    for (int i = 0; i < nums.size() - 1; i++) {
        for (int j = i + 1; j < nums.size(); j++) {
            if (nums[i] > nums[j]) {
                swap(nums[i], nums[j]);
            }
        }
    }
}

vector<int> bucketSort(vector<int>& nums) {
    vector<vector<int>> buckets(3);
    int bucketNum = 3;
    int min = nums[0], max = nums[0];
    for (int i = 0; i < nums.size(); i++) {
        min = std::min(min, nums[i]);
        max = std::max(max, nums[i]);
    }
    double bucketRange = (max - min) / bucketNum;
    for (int i = 0; i < nums.size(); i++) {
        int bucketIndex = static_cast<int>((nums[i] - min) / bucketRange);
        if (bucketIndex == bucketNum) bucketIndex--;
        buckets[bucketIndex].push_back(nums[i]);
    }

    for (auto& bucket : buckets) {
        selectSort(bucket);
    }
    vector<int> result;
    for (int i = 0; i < buckets.size(); i++) {
        for (int j = 0; j < buckets[i].size(); j++) {
            result.push_back(buckets[i][j]);
        }
    }
    return result;
}

int main(int argc, char const *argv[])
{
    vector<int> nums = {10, 3, 1, 8, 4, 7, 6};
    cout << "Before Sorted: ";
    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;
    
    vector<int> sorted = bucketSort(nums);
    cout << "After Sorted: ";
    for (int num : sorted) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}

File: sort/bucket_sort.py
def select_sort(nums: list):
    n = len(nums)
    for i in range(0, n - 1):
        for j in range(i, n):
            if nums[i] > nums[j]:
                nums[i], nums[j] = nums[j], nums[i]
def bucket_sort(nums: list):
    minV = min(nums)
    maxV = max(nums)
    bucket_num = 3
    buckets = [[], [], []]
    bucket_range = (maxV - minV) / bucket_num

    for num in nums:
        buck_index = (num - minV) / bucket_range
        print(f"buck_index is {buck_index}")
        buck_index = int(buck_index)

        if buck_index >= bucket_num:
            buck_index = bucket_num - 1
        buckets[buck_index].append(num)
    
    result = []
    for bucket in buckets:
        select_sort(bucket)
        result += bucket
    return result
if __name__ == "__main__":
    nums = [10, 3, 1, 8, 4,  7, 6]
    result = bucket_sort(nums)
    print(result)